File: frontend/src/services/waste_attendance.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { api } from './api';

export interface WasteCollector {
  id: string;
  tenant_id: string;
  resident_id?: string;
  name: string;
  phone?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface WasteAttendanceMember {
  id: string;
  attendance_id: string;
  collector_id: string;
  collector_name?: string;
  wage_amount: number;
  created_at: string;
}

export interface WasteAttendance {
  id: string;
  tenant_id: string;
  date: string;
  wage_per_person: number;
  total_wage: number;
  notes?: string;
  created_by?: string;
  members: WasteAttendanceMember[];
  created_at: string;
  updated_at: string;
}

export interface CreateCollectorPayload {
  resident_id?: string;
  name: string;
  phone?: string;
}

export interface UpdateCollectorPayload {
  resident_id?: string;
  name: string;
  phone?: string;
  is_active: boolean;
}

export interface CreateAttendancePayload {
  date: string;
  collector_ids: string[];
  wage_per_person?: number;
  notes?: string;
}

// ---------- Collectors Hooks ----------

export function useWasteCollectors(onlyActive = false) {
  return useQuery<{ data: WasteCollector[] }>({
    queryKey: ['waste-collectors', onlyActive],
    queryFn: async () => {
      const res = await api.get<{ data: WasteCollector[] }>('/waste-collectors', {
        params: { active: onlyActive ? 'true' : undefined },
      });
      return res.data;
    },
  });
}

export function useCreateWasteCollector() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: CreateCollectorPayload) => {
      const res = await api.post<WasteCollector>('/waste-collectors', payload);
      return res.data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['waste-collectors'] });
    },
  });
}

export function useUpdateWasteCollector() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, payload }: { id: string; payload: UpdateCollectorPayload }) => {
      const res = await api.put<WasteCollector>(`/waste-collectors/${id}`, payload);
      return res.data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['waste-collectors'] });
    },
  });
}

export function useDeleteWasteCollector() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      await api.delete(`/waste-collectors/${id}`);
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['waste-collectors'] });
    },
  });
}

// ---------- Attendance Hooks ----------

export function useWasteAttendance(limit = 20, offset = 0) {
  return useQuery<{ data: WasteAttendance[]; total: number }>({
    queryKey: ['waste-attendance', limit, offset],
    queryFn: async () => {
      const res = await api.get<{ data: WasteAttendance[]; total: number }>('/waste-attendance', {
        params: { limit, offset },
      });
      return res.data;
    },
  });
}

export function useCreateWasteAttendance() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: CreateAttendancePayload) => {
      const res = await api.post<WasteAttendance>('/waste-attendance', payload);
      return res.data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['waste-attendance'] });
    },
  });
}

export function useDeleteWasteAttendance() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      await api.delete(`/waste-attendance/${id}`);
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['waste-attendance'] });
    },
  });
}

File: backend/internal/domain/waste_attendance.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

// WasteCollector adalah master data pemuda / petugas pengambil sampah
type WasteCollector struct {
	ID         uuid.UUID  `json:"id"`
	TenantID   uuid.UUID  `json:"tenant_id"`
	ResidentID *uuid.UUID `json:"resident_id,omitempty"`
	Name       string     `json:"name"`
	Phone      string     `json:"phone,omitempty"`
	IsActive   bool       `json:"is_active"`
	CreatedAt  time.Time  `json:"created_at"`
	UpdatedAt  time.Time  `json:"updated_at"`
	DeletedAt  *time.Time `json:"deleted_at,omitempty"`
}

// WasteAttendanceMember adalah data petugas yang bertugas pada sesi penarikan tertentu
type WasteAttendanceMember struct {
	ID           uuid.UUID `json:"id"`
	AttendanceID uuid.UUID `json:"attendance_id"`
	CollectorID  uuid.UUID `json:"collector_id"`
	CollectorName string   `json:"collector_name,omitempty"`
	WageAmount   float64   `json:"wage_amount"`
	CreatedAt    time.Time `json:"created_at"`
}

// WasteAttendance adalah rekap absensi penarikan sampah pada suatu tanggal
type WasteAttendance struct {
	ID            uuid.UUID               `json:"id"`
	TenantID      uuid.UUID               `json:"tenant_id"`
	Date          string                  `json:"date"` // YYYY-MM-DD
	WagePerPerson float64                 `json:"wage_per_person"`
	TotalWage     float64                 `json:"total_wage"`
	Notes         string                  `json:"notes,omitempty"`
	CreatedBy     *uuid.UUID              `json:"created_by,omitempty"`
	Members       []WasteAttendanceMember `json:"members,omitempty"`
	CreatedAt     time.Time               `json:"created_at"`
	UpdatedAt     time.Time               `json:"updated_at"`
	DeletedAt     *time.Time              `json:"deleted_at,omitempty"`
}

type WasteAttendanceRepository interface {
	// Collectors
	ListCollectors(ctx context.Context, tenantID uuid.UUID, onlyActive bool) ([]WasteCollector, error)
	GetCollectorByID(ctx context.Context, tenantID, id uuid.UUID) (*WasteCollector, error)
	CreateCollector(ctx context.Context, c *WasteCollector) error
	UpdateCollector(ctx context.Context, c *WasteCollector) error
	DeleteCollector(ctx context.Context, tenantID, id uuid.UUID) error

	// Attendance
	ListAttendance(ctx context.Context, tenantID uuid.UUID, limit, offset int) ([]WasteAttendance, int, error)
	GetAttendanceByID(ctx context.Context, tenantID, id uuid.UUID) (*WasteAttendance, error)
	CreateAttendance(ctx context.Context, a *WasteAttendance) error
	DeleteAttendance(ctx context.Context, tenantID, id uuid.UUID) error
}

type WasteAttendanceUsecase interface {
	ListCollectors(ctx context.Context, tenantID uuid.UUID, onlyActive bool) ([]WasteCollector, error)
	CreateCollector(ctx context.Context, tenantID uuid.UUID, residentID *uuid.UUID, name, phone string) (*WasteCollector, error)
	UpdateCollector(ctx context.Context, tenantID, id uuid.UUID, residentID *uuid.UUID, name, phone string, isActive bool) (*WasteCollector, error)
	DeleteCollector(ctx context.Context, tenantID, id uuid.UUID) error

	ListAttendance(ctx context.Context, tenantID uuid.UUID, limit, offset int) ([]WasteAttendance, int, error)
	CreateAttendance(ctx context.Context, tenantID uuid.UUID, date string, collectorIDs []uuid.UUID, wagePerPerson float64, notes string, createdBy *uuid.UUID) (*WasteAttendance, error)
	DeleteAttendance(ctx context.Context, tenantID, id uuid.UUID) error
}
